src/main/java/com/google/jenkins/plugins/storage/DownloadStep.java

```diff
@@ -0,0 +1,428 @@
+      this.objects = Collections.unmodifiableCollection(objects);
+    }
+
+    // The Jenkins directory that will receive the objects
+    public final FilePath localDir;
+    // Objects in the cloud that need to be downloaded
+    public final Collection<StorageObject> objects;
+  }
+
+  /**
+   * Take the bucket path and return a list of objects in the cloud that match
+   * it.
+   *
+   * This will eventually handle wildcards, but for now is limited to directly
+   * specifying the object.
+   */
+  private List<StorageObject> resolveBucketPath(BucketPath bucketPath,
+      GoogleRobotCredentials credentials, String version)
+      throws IOException, ExecutorException {
+    Storage service = module.getStorageService(credentials, version);
+    Executor executor = module.newExecutor();
+
+    List<StorageObject> result = new ArrayList<StorageObject>();
+
+    // TODO(agoulti): add handling of wildcards.
+    Storage.Objects.Get obj = service.objects()
+        .get(bucketPath.getBucket(), bucketPath.getObject());
+
+    result.add(executor.execute(obj));
+
+    return result;
+  }
+
+  /**
+   * Boilerplate, see:
+   * https://wiki.jenkins-ci.org/display/JENKINS/Defining+a+new+extension+point
+   */
+  public DescriptorImpl getDescriptor() {
+    return (DescriptorImpl) checkNotNull(Hudson.getInstance())
+        .getDescriptor(getClass());
+  }
+
+  /**
+   * Descriptor for the DownloadStep
+   */
+  @Extension
+  @Symbol("googleStorageDownload")
+  public static class DescriptorImpl extends BuildStepDescriptor<Builder> {
+
+    private final UploadModule module;
+
+    public UploadModule getModule() {
+      return module;
+    }
+
+    public DescriptorImpl() {
+      this.module = new UploadModule();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getDisplayName() {
+      return Messages.Download_BuildStepDisplayName();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isApplicable(Class<? extends AbstractProject> jobType) {
+      return true;
+    }
+
+    @Override
+    public Builder newInstance(StaplerRequest req, JSONObject formData)
+        throws FormException {
+      if (Boolean.FALSE.equals(formData.remove("stripPathPrefix"))) {
+        formData.remove("pathPrefix");
+      }
+      return super.newInstance(req, formData);
+    }
+
+    /**
+     * This callback validates the {@code bucketNameWithVars} input field's
+     * values.
+     */
+    public FormValidation doCheckBucket(
+        @QueryParameter final String bucket)
+        throws IOException {
+      return ClassicUpload.DescriptorImpl.staticDoCheckBucket(bucket);
+    }
+
+    public static FormValidation doCheckLocalDirectory(
+        @QueryParameter final String localDirectory)
+        throws IOException {
+      String resolvedDir = Resolve.resolveBuiltin(localDirectory);
+      if (resolvedDir.isEmpty()) {
+        return FormValidation.error(Messages.Download_EmptyDir());
```

Comment:
Just to confirm: is it intended that we're insisting that the user enter "." to indicate that they wish to download to the root workspace directory?

Reply:
That was my original intention, but now that I think about it again, maybe it doesn't make sense.
I've removed this check, so that leaving it empty can be used to download to workplace.